Refactor `File\RenameUpload`

gsteel opened the pull request:
Includes changes in #223

This patch further simplifies the filter, making use of `FileInformation` to handle the different possible acceptable inputs, and helps get rid of the awkward additional PSR dependencies by only returning the file path of the uploaded file.

Adds the `MoveUploadedFile` utility as a constructor dependency, requiring a factory setup with the plugin manager.

Expands `MoveUploadedFile` to take care of basic error handling

Reverts `target_directory` option and semantics back to `target`, hopefully, for a simpler transition.

I'd appreciate review and input from @ramchale, @marcelthole and @froschdesign 👍

marcelthole committed:
Refactor RenameUpload filter to implement FlterInterface

Signed-off-by: Marcel Thole <marcel@marcelthole.de>

gsteel committed:
Further refactoring of `RenameUpload`

- Renames `MoveUploadedFileMover` to `MoveUploadedFile`
- Re-locates error handling for moving uploaded files to the implementation, upcasting native warnings/errors to throwables
- Further simplifies `RenameUpload` to operate on a single file
- Makes use of the `FileInformation` class to remove internal dependencies on stream factories etc.
- Reverts the option name `target_directory` back to `target` and disambiguates it internally to a directory and filename component

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Introduce a factory for the RenameUpload filter

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Suppress Internal class usage issues in tests

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Update documentation examples for `RenameUpload`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Remove the now unused `AbstractFilter`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Update migration guide with rename-upload changes

Signed-off-by: George Steel <george@net-glue.co.uk>

froschdesign commented on src/File/RenameUploadFactory.php:
```diff
@@ -0,0 +1,22 @@
+        /**
+         * @psalm-suppress MixedArgumentTypeCoercion Runtime validation of options is not worth it.
+         */
+        return new RenameUpload($options ?? [], new MoveUploadedFile());
```
Have I overlooked something or why is the factory needed?

gsteel replied:
Arguably, we could inline the `MoveUploadedFile` class. It was extracted so that we could stub PHP's `move_uploaded_file` function, but now the default implementation checks `is_uploaded_file` _(and runs `rename` if its not)_, it feels less important that its a stand-alone collaborator. The only other benefit of keeping it external is that it now also runs an error handler to upcast PHP errors/warnings to exceptions, but we could also just `new` it in the constructor and drop the factory completely.

What do you think?

froschdesign replied:
> …but we could also just `new` it in the constructor and drop the factory completely.

It is already there:

```php
$this->moveUploadedFile = $moveUploadedFile ?? new MoveUploadedFile();
```

https://github.com/laminas/laminas-filter/pull/239/files#diff-94631e5208b9361631c5713c4af740513be7608554440d9130f6e8f633498bd5R62

gsteel replied:
Thanks @froschdesign - I've dropped the factory now 👍

marcelthole commented on src/File/MoveUploadedFile.php:
```diff
@@ -0,0 +1,47 @@
+            throw new ErrorException($message, $errorNumber, $errorNumber, $file, $line);
+        });
+
+        if (! is_uploaded_file($sourceFile)) {
```
Isn't this the security feature of `move_uploaded_file` that will be disabled with this condition?
This Filter should be used for uploaded files only. If someone hijack the $_FILES object this security mechanism will not work anymore. 

You need this only for the Tests and there is already the MoveUploadedFileInterface. Would prefer to use here the move_uploaded_file method and use the rename Handler only for the tests

gsteel replied:
Good point, running `is_uploaded_file` in the filter and rejecting input upfront would cover this. We could then continue to use the existing (And internal) `MoveUploadedFile` implementation in the knowledge that the conditional there is only used in test scenarios.

gsteel replied:
Brains aren't working well today - thanks for the review @marcelthole - I've reverted closer to your original work 👍

gsteel committed:
Drop the factory for `RenameUpload`

The factory is unnecessary because the default implementation is "newed" in the constructor when null.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Correct default `MoveUploadedFile` implementation

Checking for an uploaded file and falling back to a regular `rename` is a potential security issue considering we should only be processing uploads with this filter, so, the default implementation now only wraps `move_uploaded_file` and a test implementation is added using rename.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
If we can get this over the line before the end of the month, we can vote on releasing v3 at the next TSC meet in May.

ramchale reviewed (approved):
Looks like a good simplification to me

froschdesign commented on docs/book/v3/file.md:
```diff
@@ -166,6 +166,8 @@ echo $filter->filter('fileB.txt');
 
 `Laminas\Filter\File\RenameUpload` can be used to rename or move an uploaded file to a new path.
 
+It will accept an array representing a single [PHP file upload](https://www.php.net/manual/features.file-upload.post-method.php), a PSR7 [`UploadedFileInterface`](https://www.php-fig.org/psr/psr-7/#36-psrhttpmessageuploadedfileinterface), or a string representing the file path of an uploaded file.
```
An unordered list for the types would give a better overview.

froschdesign commented on docs/book/v3/file.md:
```diff
@@ -203,13 +197,11 @@ The following set of options are supported:
 ### Usage Examples
 
-Move all filtered files to a different directory:
+Move all filtered files to a different directory in a Laminas MVC application:
```
I do not understand the addition. Does not this also work in simple scripts without laminas-mvc?

gsteel replied:
I've removed that addition and changed one of the examples to use `$_FILES` directly.

Also, dropped the second, nullable arg to the constructor in the examples which is unnecessary and probably will just cause confusion for users

froschdesign replied:
Thank you! 👍🏻 

In principle, the components should always be described as they function on their own - if this is possible, of course.
The explanations for Mezzio, laminas-mvc or Slim & Co. must be described separately under a separate navigation point / page.

gsteel committed:
Improve docs for `RenameUpload`

- Clearly state the acceptable inputs by listing them in an un-ordered list
- Remove the potentially confusing 2nd argument which is unnecessary in general usage.
- Alter examples so that at least one of them uses PHP's `$_FILES` array directly instead of via MVC's Request object.

Signed-off-by: George Steel <george@net-glue.co.uk>
